### PR TITLE
feat: can set a maximum of selected items [maxSelectedItems] when mult…

### PR DESCRIPTION
**Pull request by benjaminlefevre**
…iple selection is activated

https://github.com/ng-select/ng-select/issues/181

### CODE REVIEW

File: src/ng-select/items-list.ts [hunk 1]
```diff
@@ -188,4 +188,8 @@ export class ItemsList {
     private get _lastSelectedItem() {
         return this._selected[this._selected.length - 1];
     }
+
+    isMaximumNumberOfSelectedItemsReached(): boolean {
```
Comment (anjmao) [inline src/ng-select/items-list.ts]: Move this public method before all private methods.

File: src/ng-select/ng-select.component.ts [hunk 2]
```diff
@@ -269,7 +274,7 @@ export class NgSelectComponent implements OnInit, OnDestroy, OnChanges, AfterVie
     }
 
     open() {
-        if (this.isDisabled || this.isOpen) {
+        if (this.isDisabled || this.isOpen || this.isMaximumNumberOfSelectedItemsReaches()) {
```
Comment (anjmao) [inline src/ng-select/ng-select.component.ts]: You can remove that method and use one prod itemsList directly 
```js
if (this.isDisabled || this.isOpen || this.itemsList.isMaximumNumberOfSelectedItemsReaches()) {
```

Comment (andreasfranz) [reply src/ng-select/ng-select.component.ts]: I think in this case it should be `this.itemsList.isMaximumNumberOfSelectedItemsReached()` then

**Comment by benjaminlefevre** [comment]
I will fix the issues this evening after work, thank you for the review

**Comment by andreasfranz** [comment]
Hi @anjmao,
I also thought of such a feature recently, because I would need it for my project too. I think the suggested solution of @benjaminlefevre  as a certain drawback: From a user's point of view it hard to recognize why the select box is suddenly disabled (without any further hint in the user interface).

As using a single select behaves differently: Once you select a new option is just replaces the previous one. Maybe this could also be an option, just replace the last item in the list with the newly selected. I am not sure if this is a better behavior than disabling the select box.
What do you think?

File: src/ng-select/items-list.ts (hunk 1 above)
Comment (varnastadeus) [reply src/ng-select/items-list.ts]: Also I would suggest renaming this method to less verbose, eg.: `maxItemsSelected` or similar

**Comment by anjmao** [comment]
@andreasfranz replacing last selected item can look a bit strange and user always can remove item from the list so he have a choose. I think current implementation is good and it would be even less clear if dropdown is opened but you can't select items because max items are reached. Best approach would be to add some kind of hint (label) for user that he can select only x items.

**Comment by benjaminlefevre** [comment]
I made the fixes.

And yes, I made a choice, which seemed to me the most natural: not to replace the last item, but to block the selection of items. 
It would be nice to find a visual trick to indicate the max is reached, I thought of a tooltip but what about the internationalization.... I think it could be done afterwards, since for the moment this is a functionality which will be not used a lot.

**Pull request merged**